### PR TITLE
Support Dependent Identity Data and "After" Param in Saas Configs [#291]

### LINKED ISSUE #291: [Saas Connector Engine][Fidesops Core] Configure "After" Param for Saas Configs and Dependent Identities
❗ Blocks #256 

# What
Allow an endpoint with identity data to be specified that it needs to wait on all incoming inputs before visiting (for saas configs).  

# Why

Currently, when we build the graph of all nodes we need to visit, we create a root node, and its children are any nodes with identity data.    Other nodes wait for all of their upstream nodes to have been visited, but nodes with identity data are treated differently.  We need to be able to specify that an identity node should be sometimes be visited only after every other input is fulfilled.

## Use case

A saas connector needs a `project_id` and an `organization_id` from one endpoint, and needs to use the identity data as a query param.  We can't attempt to visit this endpoint first, because we don't yet have the `project_id` and `organization_id`.

## Alternatives considered

- There's an `after` param that can be specified on a collection, look into using that.  On a quick glance though, it seems it's a little more of an involved fix.
- We could also use a postprocessor to filter the data with the identity after the fact, but  I don't view this as a viable solution because of the enormous amount of extra queries we'd need to make.

### PR DESCRIPTION
**Pull request by pattisdr**
❗ Dependent on #299 

# Purpose

In Fidesops, we can specify that a collection should be forced to run after another collection.  Likewise, we need to be able to specify that a Saas Config endpoint is run "after" another collection or endpoint. This is to support the use case where an endpoint with an identity data needs to wait for a response from another endpoint before running.

Furthermore, as a follow-up to #299, where we allowed inputs into a collection to be grouped together, we need to support grouping identity data as well.  For example, a saas config endpoint needs to wait for two references from one endpoint, and also needs identity data before proceeding. These three pieces of information need to stay grouped together.

# Changes
- Support the "after" param on a Saas Config endpoint
- If an identity field is listed alongside collection references in `grouped_inputs`, combine the identity data with the other grouped_inputs in `GraphTask.pre_process_input_data` before passing that data onto the endpoint.

# Checklist

- [x] Applicable documentation updated (guides, quickstart, postman collections, tutorial, fidesdemo, [database diagram](https://github.com/ethyca/fidesops/blob/main/docs/fidesops/docs/development/update_erd_diagram.md))
- [x] Good unit test/integration test coverage **(Note: end-to-end testing will be completed with Sentry #256 )**
- [ ] This PR contains a DB migration. If checked, the reviewer should confirm with the author that the [down_revision correctly references the previous migration](https://ethyca.github.io/fidesops/development/contributing_details/#alembic-migrations) before merging
- [ ] The `Run Unsafe PR Checks` label has been applied, and checks have passed, if this PR touches any external services

# Ticket

Fixes #291

### CODE REVIEW

**Comment by pattisdr** [comment]
Addressing test failures ^

**Comment by pattisdr** [comment]
Rebasing ^

**Comment by pattisdr** [comment]
Getting this up-to-date after saas config headers merge ^

File: src/fidesops/schemas/saas/saas_config.py [hunk 1]
```diff
@@ -214,6 +222,10 @@ def get_graph(self) -> Dataset:
                         name=endpoint.name,
                         fields=fields,
                         grouped_inputs=grouped_inputs,
+                        after={
+                            CollectionAddress(*s.split("."))
+                            for s in endpoint.after or []
```
Comment (seanpreston) [inline src/fidesops/schemas/saas/saas_config.py:227]: `endpoint.after` will default to `[]` so we don't need `or []` unless you think it helps readability

File: src/fidesops/task/graph_task.py [hunk 2]
```diff
@@ -128,6 +128,15 @@ def grouped_fields(self) -> Set[Optional[str]]:
...
         return self.traversal_node.node.collection.grouped_inputs or set()
 
+    @property
+    def dependent_identity_fields(self) -> bool:
+        """If the current collection needs inputs from other collections, in addition to its seed data."""
+        collection = self.traversal_node.node.collection
+        grouped_fields = [
+            collection.field(FieldPath(field)) for field in self.grouped_fields
+        ]
+        return any(field.identity for field in grouped_fields)
```
Comment (seanpreston) [inline src/fidesops/task/graph_task.py:138]: We could save some iterations here using a for loop:

```
for field in grouped_fields:
    if collection.field(FieldPath(field)).identity:
        return True
```

then we also wouldn't need the overlap of `grouped_fields` and `self.grouped_fields`, where one is a different format.

File: src/fidesops/task/graph_task.py [hunk 3]
```diff
@@ -168,6 +177,27 @@ def can_write_data(self) -> bool:
         connection_config: ConnectionConfig = self.connector.configuration
         return connection_config.access == AccessLevel.write
 
+    def _combine_seed_data(
+        self,
+        *data: List[Row],
+        grouped_data: Dict[str, Any],
+        dependent_field_mappings: COLLECTION_FIELD_PATH_MAP,
+    ) -> Dict[str, Any]:
+        """Combine the seed data with the other dependent inputs. This is used when seed data into a collection requires
```
Comment (seanpreston) [inline src/fidesops/task/graph_task.py:186]: Should read `...when seeding data...`

File: tests/task/test_graph_task.py [hunk 4]
```diff
@@ -249,15 +249,14 @@ def test_pre_process_input_conversation_collection(
         }
 
     def test_pre_process_input_data_group_dependent_fields(self):
-        """Test processing inputs where fields have been marked as dependent"""
+        """Test processing inputs where fields have been marked as dependent
+        Identity data has also been marked as being dependent.
```
Comment (seanpreston) [inline tests/task/test_graph_task.py:253]: Should this read `Test processing inputs where identity data has also been marked as being dependent.`?

**Review by seanpreston — APPROVED** [review]
Awesome work @pattisdr — just a couple of minor suggestions here.

File: src/fidesops/task/graph_task.py (hunk 3 above)
Comment (pattisdr) [reply src/fidesops/task/graph_task.py:186]: hm I meant it as a noun, maybe adding `the` will help? This is used when the seed data in a collection requires inputs from another collection to generate subsequent queries.

File: tests/task/test_graph_task.py (hunk 4 above)
Comment (pattisdr) [reply tests/task/test_graph_task.py:253]: Well it's testing both, that reference fields and an identity field have all been marked together as a group, I'll rephrase!

File: src/fidesops/task/graph_task.py (hunk 2 above)
Comment (pattisdr) [reply src/fidesops/task/graph_task.py:138]: Good idea

File: src/fidesops/schemas/saas/saas_config.py (hunk 1 above)
Comment (pattisdr) [reply src/fidesops/schemas/saas/saas_config.py]: Ah I added this for mypy, I think if I remove the Optional from endpoint.after it should work.

**Comment by pattisdr** [comment]
Thanks for the review @seanpreston, ready for another pass!

**Pull request merged**